scanpy/get/_aggregated.py

```diff
@@ -0,0 +1,437 @@
+    If none of `layer`, `obsm`, or `varm` are passed in, `X` will be used for aggregation data.
+    If `func` only has length 1 or is just an `AggType`, then aggregation data is written to `X`.
+    Otherwise, it is written to `layers` or `xxxm` as appropriate for the dimensions of the aggregation data.
+
+    Params
+    ------
+    adata
+        :class:`~anndata.AnnData` to be aggregated.
+    by
+        Key of the column to be grouped-by.
+    func
+        How to aggregate.
+    dim
+        Axis on which to find group by column.
+    weight_key
+        Key of the `dim` containing weights for a weighted sum aggregation.
+    key_set
+        Subset of dim on which to filter.
+    dof
+        Degrees of freedom for variance. Defaults to 1.
+    layer
+        If not None, key for aggregation data.
+    obsm
+        If not None, key for aggregation data.
+    varm
+        If not None, key for aggregation data.
+
+    Returns
+    -------
+    Aggregated :class:`~anndata.AnnData`.
+    """
+    data = adata.X
+    write_to_xxxm = None
+    if varm is not None:
+        data = adata.varm[varm]
+        write_to_xxxm = True  # the data will have to be transposed so this is accurate
+    elif obsm is not None:
+        data = adata.obsm[obsm]
+        write_to_xxxm = True
+    elif layer is not None:
+        data = adata.layers[layer]
+        if dim == 'var':
+            data = data.T
+    elif (
+        dim == 'var'
+    ):  # i.e., all of `varm`, `obsm`, `layers` are None so we use `X` which must be transposed
+        data = data.T
+    return aggregated(
+        data,
+        groupby_df=getattr(adata, dim),
+        dim=dim,
+        by=by,
+        write_to_xxxm=write_to_xxxm,
+        no_groupby_df=getattr(adata, 'var' if dim == 'obs' else 'obs'),
+        weight_key=weight_key,
+        key_set=key_set,
+        func=func,
+        dof=dof,
+    )
+
+
+@aggregated.register(np.ndarray)
+@aggregated.register(spmatrix)
+def aggregated_from_array(
+    data,
+    groupby_df: pd.DataFrame,
+    func: Union[AggType, List[AggType]],
+    dim: str,
+    by: str,
+    write_to_xxxm: bool,
+    no_groupby_df: pd.DataFrame,
+    weight_key: Optional[str] = None,
+    key_set: Optional[Iterable[str]] = None,
+    dof: int = 1,
+) -> AnnData:
+    """Aggregate data based on one of the columns of one of a `~pd.DataFrame`."""
+    groupby = Aggregate(
+        groupby=groupby_df[by],
+        data=data,
+        weight=groupby_df[weight_key] if weight_key is not None else None,
+        key_set=key_set,
+    )
+    # groupby df is put in `obs`, nongroupby in `var` to be transposed later as appropriate
+    obs_var_dict = {'obs': _df_grouped(groupby_df, by, key_set), 'var': no_groupby_df}
+    data_dict = {
+        'layers': {},
+        'X': None,
+        'obsm': {},
+    }
+    func_set = func
+    write_key = 'obsm' if write_to_xxxm else 'layers'
+    if not isinstance(func, list):
+        func_set = [func]
+    func_set = set(func_set)
```

Comment:
Parameter signatures shouldn’t contain concrete container classes (except for `tuple`), and neither should you use them in typechecks. Be lenient in what you accept (in this case a string or an iterable of strings) and strict in what you return (`list[X]` is fine as a return type). Therefore the parameter type should be `Iterable[AggType] | ArgType` here and you can do that simply like this:

```suggestion
    funcs = set([func] if isinstance(func, str) else func)
    write_key = 'obsm' if write_to_xxxm else 'layers'
```

afterwards you should validate that all entries of `funcs` are handled, e.g. by doing

```py
if unknown := funcs - set(get_args(AggType)):
    raise ValueError(f'… {unknown} …')
``` 

(see [`typing.get_args`](https://docs.python.org/3/library/typing.html#typing.get_args))